driver/utils/unicode_converter.h

```diff
@@ -0,0 +1,722 @@
+    const bool ensure_src_signature = true,
+    const bool trim_dest_signature = true
+);
+
+template <typename SourceCharType, typename DestinationCharType, typename PivotCharType>
+inline void convertEncoding(
+    UnicodeConverter & src_converter, const std::basic_string<SourceCharType> & src,
+    std::basic_string<PivotCharType> & pivot,
+    UnicodeConverter & dest_converter, std::basic_string<DestinationCharType> & dest,
+    const bool ensure_src_signature = true,
+    const bool trim_dest_signature = true
+) {
+    return convertEncoding(src_converter, make_string_view(src), pivot, dest_converter, dest, ensure_src_signature, trim_dest_signature);
+}
+
+class UnicodeConverter {
+public:
+    explicit inline UnicodeConverter(const std::string & encoding);
+    inline ~UnicodeConverter();
+
+    inline const std::size_t getEncodedMinCharSize() const;
+
+    template <typename CharType>
+    inline std::basic_string_view<CharType> consumeEncodedSignature(const std::basic_string_view<CharType> & str) const;
+
+    template <typename CharType>
+    inline std::basic_string_view<CharType> consumePivotSignature(const std::basic_string_view<CharType> & str) const;
+
+    template <typename CharType>
+    inline std::size_t consumeEncodedSignatureInPlace(std::basic_string<CharType> & str) const;
+
+    template <typename CharType>
+    inline std::size_t consumePivotSignatureInPlace(std::basic_string<CharType> & str) const;
+
+    template <typename CharType, typename PivotCharType>
+    inline void convertToPivot(
+        const std::basic_string_view<CharType> & encoded,
+        std::basic_string<PivotCharType> & pivot,
+        const bool ensure_encoded_signature,
+        const bool trim_pivot_signature
+    );
+
+    template <typename CharType, typename PivotCharType>
+    inline void convertFromPivot(
+        const std::basic_string_view<PivotCharType> & pivot,
+        std::basic_string<CharType> & encoded,
+        const bool ensure_pivot_signature,
+        const bool trim_encoded_signature
+    );
+
+    template <typename SourceCharType, typename DestinationCharType, typename PivotCharType>
+    friend inline void convertEncoding(
+        UnicodeConverter & src_converter, const std::basic_string_view<SourceCharType> & src,
+        std::basic_string<PivotCharType> & pivot,
+        UnicodeConverter & dest_converter, std::basic_string<DestinationCharType> & dest,
+        const bool ensure_src_signature,
+        const bool trim_dest_signature
+    );
+
+private:
+    UConverter * converter_ = nullptr;
+
+    std::string encoded_signature_to_prepend_;
+    std::vector<std::string> encoded_signatures_to_trim_;
+    std::size_t encoded_signatures_to_trim_max_size_ = 0;
+
+    std::string pivot_signature_to_prepend_;
+    std::vector<std::string> pivot_signatures_to_trim_;
+    std::size_t pivot_signatures_to_trim_max_size_ = 0;
+};
+
+inline UnicodeConverter::UnicodeConverter(const std::string & encoding) {
```

Comment:
I really do not think that inlining function this big is a good idea: 
* first of all, it is so less likely to be actually inlined due to the size
* even if it is inlined, it will generate HUGE amount of extra code in many translation units.
* even if it is inlined, there will be no practical benefit performance-wise from inlining, since that would save you what? My estimate is approx 0.1-0.2% of execution time.
* that also leads to implementation details like `sameEncoding`, `make_raw_str`, `consumeSignature`, etc. leak to the common namespace.
* due to `UnicodeConverter` being header-only, it's implementation details are spilled all over the source base of the entire project. That produces project so coupled so highly that (as it looks like right now) any change in ICU (or attempt to replace it with something else) will ripple through the whole project.

Reply:
I will move everything that can be placed in .cpp files, there. However, inlining here is more of an attempt to help the compiler to do deeper static analysis during optimization. In fact, most of the modern compilers will still inline this code, even if it is put in .cpp, thanks to LTO. So, size of the binary, and other things will be the same. Maybe slightly better compilation time, however, also slightly higher risks of slightly worse optimizations, if compiler+linker are not fresh enough.

Reply:
Done.